### PR TITLE
Bugfix gradient formula for index_reduce('prod') + separate out sample_inputs for index_reduce

### PR DESCRIPTION
**Pull request by mikaylagawarecki**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #76996
* __->__ #77382
* #76997
* #76296

- Applies the same backward fix for `index_reduce(prod)` as was applied to `scatter_reduce(prod)` in #76523 (gradients of 0s in self/source should not be zeroed out)
- Adds a separate test for even propagation of gradients for min/max for scatter/index_reduce as this case will fail on gradgradcheck
- Separates out sample_inputs_function for index_reduce [as per this comment](https://github.com/pytorch/pytorch/pull/75981#discussion_r860578084)

### CODE REVIEW

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -7679,17 +7706,6 @@ def _gather(shape, index_dim, max_indices):
...
                           kwargs={'include_self': True})
 
-
-    # Check that gradients are evenly distributed for amax/amin
-    if requires_grad and (reduce in ['amax', 'amin']):
-        input = _tensor((M, S))
-        src = input.clone().detach_().requires_grad_(requires_grad)
-        idx = torch.arange(S).expand((M, S)).to(dtype=torch.long, device=device)
-
-        yield SampleInput(_tensor((M, S)),
```
Comment (mikaylagawarecki) [inline torch/testing/_internal/common_methods_invocations.py:7689]: There was a bug here where I passed a new tensor rather than input so the gradgrad failure wasn't caught.

**Review by cpuhrsch — APPROVED** [review]
